aten/src/ATen/TensorIndexing.h

```diff
@@ -363,19 +363,36 @@ static inline void copy_to(const Tensor& dst, const Tensor& src) {
 }
 
 // See NOTE [ Setting `disable_slice_optimization` when calling C++ tensor indexing functions from Python ]
+// prev_tensor_dim_ptr is used to keep track of the dimension of the previously
+// passed tensor.  Must be -1 when not set.
+// https://github.com/pytorch/pytorch/issues/71673
 static inline Tensor handleDimInMultiDimIndexing(
     const Tensor& prev_dim_result,
     const Tensor& original_tensor,
     const TensorIndex& index,
     int64_t* dim_ptr,
     int64_t* specified_dims_ptr,
+    int64_t* prev_tensor_dim_ptr,
```

Comment:
Could you please insert a detailed comment describing why this fixes the issue?

Reply:
i actually have no idea why this works. the way i came up with this is by observing the behavior of different tests cases. iirc i noticed that the only thing that's different between the equivalent cases (`t[m,0]` vs `t[:,:,0][m]` for a 2d `m`) is the dimension pointer in the integer case, so i changed it and it worked for the original repro (the case i've just mentioned). then i added a bunch of other cases and it worked for all of them. (i also checked everything against numpy).

i'm still not sure whether this is a proper way to do it, but i have no other ideas. note that i have tests for 2d, 3d, 1d (well, this one is skipped due to an `if` in the above function, which is kinda sus), as well as a test where the tensor argument is not the very first one. that said, i wonder if we should add more tests. e.g., maybe with a slightly different shape for input and mask (perhaps i just got lucky with shapes/strides/whatever).

ed wrote an explainer here (this is linked in the issue):
https://dev-discuss.pytorch.org/t/how-does-advanced-indexing-work-when-i-combine-a-tensor-and-a-single-index/558

but the general pattern is as follows:

- for a single tensor: `t[m,i] == t[:,:,i][m]` where the number of `: == m.dim`
- for multiple tensors: `t[m,i,n,j] == t[:,i,:,j][m,n]` where `m` and `n` are 1d tensors, and `i` and `j` are integers

one of the problems is that coming up with test cases for this is pretty annoying because you need to make sure that the shapes match. lmk if you have any other ideas for things that are not covered here. perhaps we could try introducing other indexing type in between a tensor and integer: `t[m,???,i]`(?)

Reply:
yeah, before this can go, i think we definitely need to test a combination of all these with a tensor index somewhere and figure out why the 1d tensor case fails with this fix (without the `if`):

```
  if (index.is_integer()) {
  } else if (index.is_slice()) {
  } else if (index.is_ellipsis()) {
  } else if (index.is_none()) {
  } else if (index.is_boolean()) {
  } else if (index.is_tensor()) {
```

Reply:
heads up: i think i have a way to fix this by adjusting `dim_ptr` directly, will do more tests and push shortly if it works. but the general idea would be the same (incrementing `dim_ptr` by the number of dimensions of the previously passed tensor index).

Reply:
I think you need this logic from `count_specified_dimensions`:
```
      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
        count += tensor.dim();
      } else {
        count++;
      }
```

Reply:
nice, this seems to work! the funny thing is i actually looked at count_specified_dimensions for issue 73236 not that long ago, but failed to make the connection here... let's see what ci thinks.

Reply:
i'll try adding a few more tests maybe to make sure.

Reply:
@peterbell10 case 4: bool fails here (different shapes): https://github.com/pytorch/pytorch/pull/75170/commits/5107071066ba846ff98706fd04726f7388488d62

any ideas how to fix this properly without changing a lot of code?